### DISCUSSION TITLE
Write objects to the pipeline

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [SamErde/PSPreworkout#4] [General]
I looked at the code for `Show-LoadedAssemblies`. This is a great idea. But your functions severely restricts how I can use this code. You are assuming I want to use one of the gridview commands. What if I want to see the output formatted as a list? What if I want to export it to a JSON file? I can't do anything with this command other than what you force me to do. Functions should write objects to the pipeline. Let the user decide how they want to consume the output.   

This function should really be `Get-LoadedAssemblies` which wraps around this code:

```powershell
 [System.AppDomain]::CurrentDomain.GetAssemblies() |
        Where-Object -FilterScript { $_.Location } |
            Sort-Object -Property FullName 
```

By the way, I would suggest including the computername and powerShell version as additional properties. If I export this data and look at later, it would be helpful to know what computer I was looking at. Give the user flexibility in how they might use the output. Don't assume they will only look at the result on the screen once.

**Comment by SamErde** [comment]
Thanks, I like this suggestion a lot and will incorporate it. The initial idea was simply a way to remind myself of what assemblies are available, and your input (er, output?) 😊 will make the results even more useful! I appreciate the feedback!

**Comment by jdhitsolutions** [comment]
If you know that you, personally, always want output formatted a specific way, then create a simple wrapper script or function around the module command and call that. Module commands should write one type of rich object to the pipeline. No formatting.

**Comment by SamErde** [comment]
That actually is why I chose the verb `Show` instead of `Get`. Still, I do like the points that you called out.

**Reply by jdhitsolutions** [reply]
I assumed that was the reason, but then you should have a simple `Get` version.

**Comment by SamErde** [comment]
Or just the `Get` version with an `OutGrid` parameter?

**Reply by jdhitsolutions** [reply]
I would say that breaks the pattern. `Get` should *get* something and write it to the pipeline. The question is, does formatted output need to be part of this module? You can't anticipate what commands a user might want to use for formatting or what version of PowerShell they might be using. And what if some new fancy formatting tool becomes available? I'd suggest write the `Get` command and call it a day. You can show help examples where the command is piped to Out-GridView.  

And if your output has a unique type name, you can create a default formatted view or even define alternate views.

**Comment by SamErde** [comment]
Hmmm, I already added the `GridView` parameter as a non-default switch, but you've given me good food for thought regarding patterns.

**Reply by jdhitsolutions** [reply]
Native module commands should write native objects to the pipeline. You have no way of knowing *how* a user will consume the command's output, nor do you want to put them into a corner. Technically, the `Show` verb is to display resources or items that might otherwise be hidden.

**Reply by jdhitsolutions** [reply]
Here's another example: `Get-DnsServerCache` retrieves cache settings, while `Show-DnsServerCache` gets the actual cache entries. In my opinion, the first command should have been called `Get-DnsServerCacheSettings` and the second command `Get-DsnServerCache`. But nobody ever asks for my opinion. 😄 

Naming things is as much an art as anything. I think the name should be meaningful and concise.

**Comment by SamErde** [comment]
OK, I took your advice from `Get-Types` and had fun with this one. `Get` is simply getting the objects and now `Show-LoadedAssembly` is handling presentation. I created custom objects so all of the metadata isn't locked in the "FullName" attribute.
